### ISSUE TITLE
Port selected dynamically in WSL2 not available in Windows via localhost

### ISSUE THREAD

**Issue by nick-beer** [microsoft/WSL#9805]
### Windows Version

10.0.22000.1696

### WSL Version

1.1.3.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.90.1

### Distro Version

Ubuntu 20.04.5

### Repro Steps

In WSL:
```bash
dotnet new web -n dynamic-test
cd dynamic-test
dotnet run --urls http://*:0
```
In Windows:
```Powershell
Invoke-WebRequest http://localhost:<dynamic_port_from_wsl>
```

### Expected Behavior

The `Invoke-WebRequest` call completes successfully, with "Hello World!" as the response.  

I get this expected behavior if any of the following are true:
- I specify a static port (`--urls http://*:41215`)
- I'm using WSL 1
- I'm on a physical machine (not a DevBox - see "Important Note")

### Actual Behavior

The `Invoke-WebRequest` call fails with `Invoke-WebRequest: No connection could be made because the target machine actively refused it.`

## Important Note
We use [Microsoft DevBox](https://azure.microsoft.com/en-ca/products/dev-box/) at our company, and this issue only seems to reproduce on instances of DevBox.

I've captures strace logs and run networking.bat.  I'll email those to InsideSupport as instructed.

**Comment by nick-beer** [comment]
I just verified that this isn't just a dotnet/kestrel issue.  Opening a dynamic port as follows shows the same issue:
```bash
# Requires php 8: https://php.watch/versions/8.0/built-in-server-port-zero
sudo apt-get install php-cli
php -S 0.0.0.0:0
```

**Comment by OneBlue** [comment]
@nick-beer: The issue is mentioning both WSL1 and WSL2. Which one are you using ?

**Comment by nick-beer** [comment]
@OneBlue - sorry for the confusion.  I'm using WSL2.  I was trying to indicate that everything works as expected if I use WSL1.

**Comment by nick-beer** [comment]
I've attached the logs.  Here's what I did:
1. `wsl --shutdown` (admin terminal)
2. .\collect-wsl-logs.ps1 (admin terminal)
3. Launched ubuntu 22.04 in wsl2 (new, non-admin terminal)
4. ran `php -S 0.0.0.0:0`
5. From powershell on Windows, ran `Invoke-WebRequest http://localhost:33141` - where 33141 is the dynamic port selected by the previous `php` call (new, non-admin terminal)

![image](https://user-images.githubusercontent.com/1509215/226787016-171ad993-3148-4ed3-bf40-c38a6963d0e6.png)


[WslLogs-2023-03-21_21-27-06.zip](https://github.com/microsoft/WSL/files/11035650/WslLogs-2023-03-21_21-27-06.zip)

**Comment by nick-beer** [comment]
@benhillis - here you go!
![image](https://user-images.githubusercontent.com/1509215/227638862-d5fce7a4-7493-4b2a-9110-d1b61d4a27a0.png)
[wsl_cmd_line.txt](https://github.com/microsoft/WSL/files/11066602/wsl_cmd_line.txt)

**Comment by benhillis** [comment]
@nick-beer thanks! Was looking for the wslhost that's running "-mode 2", because I didn't see any telemetry coming from it. When it detects sets up a relay there should be some events logged.

**Comment by nick-beer** [comment]
@benhillis - interesting.  Is there something else I can do to provide more information?

**Comment by olivierchatry** [comment]
Still having issue with latest WSL ( 1.1.3 ), using minikube. I had to rever to 1.0.3 ( latest working WSL for me ). If I can help in anyway, tell me what you need.

**Comment by jiasli** [comment]
This issue is affecting multiple [Azure CLI](https://github.com/Azure/azure-cli) users: https://github.com/Azure/azure-cli/issues/26180

Azure CLI and its underlying library [MSAL](https://github.com/AzureAD/microsoft-authentication-library-for-python) open a dynamic port to listen to the auth code in WSL, but the request made by host Windows's web browser can not reach WSL and fails with `ERR_CONNECTION_REFUSED`.

**Comment by pmartincic** [comment]
@jiasli, as others have noted on the thread you cited it should be fixed in 1.2.1. 

And looking at the logs I think this is the same issue as what Ben fixed in 1.2.1
